File: apps/ios/Packages/TrackEvolutionKit/Sources/TrackEvolutionKit/Offline/OfflinePatch.swift
import Foundation
import GRDB

// Optimistic patching: a queued mutation edits the cached responses it affects, so
// the UI shows the change immediately and every view agrees — the dashboard, the
// event page and the track page are three different cache entries describing the
// same rows.
//
// Patches are best-effort *display* state. A missing cache entry is skipped, never
// created: the queue is the source of truth until it flushes. Every patch is
// idempotent, because `reapplyQueue` runs them again over a fresh server response.
//
// Unlike the web mirror, which surgically edits plain JSON, this decodes into the
// typed models and re-encodes. That's only safe because the golden contract test
// proves those models round-trip losslessly — a field this layer doesn't know about
// would otherwise be silently dropped on every patch.

extension OfflineStore {
    /// A create's request body, in the server's field names.
    private struct EventBody: Decodable {
        var track_id: Int?
        var track_name: String?
        var start_date: String?
        var days: Int?
        var club: String?
        var run_group: String?
        var car: String?
        var notes: String?
        var conditions: Conditions?
        var temp_f: Int?
        var checklist: [ChecklistItem]?
        var best_time_ms: Int?
        var track_hours: Double?
    }

    private struct SessionBody: Decodable {
        var label: String?
        var notes: String?
        var laps: [Double]?
        var trace: [TracePoint]?
        var channels: SessionChannels?
    }

    private struct LapsBody: Decodable {
        var laps: [Double]?
    }

    private struct TrackBody: Decodable {
        var name: String?
        var notes: String?
        var goal_ms: Int?
    }

    /// Local ids for rows the server hasn't seen. Derived from the queue id so a
    /// re-apply produces the same ids and dedupes instead of duplicating.
    private static func tempLapId(seed: Int, lapNum: Int) -> Int {
        -(abs(seed) * 1000 + lapNum)
    }

    static func applyLocal(_ item: QueuedWrite, _ db: Database) throws {
        let s = item.path.split(separator: "/").map(String.init)
        let method = item.method

        if method == "POST", s == ["events"] {
            try createEvent(item, db)
        } else if method == "PUT", s.count == 2, s[0] == "events" {
            try updateEvent(id: s[1], item, db)
        } else if method == "DELETE", s.count == 2, s[0] == "events" {
            try deleteEvent(id: s[1], db)
        } else if s.count == 4, s[0] == "events", s[2] == "setups", method == "PUT" || method == "DELETE" {
            try patchSetup(eventId: s[1], day: s[3], item, db)
        } else if method == "POST", s.count == 3, s[0] == "events", s[2] == "sessions" {
            try addSession(eventId: s[1], item, db)
        } else if s.count == 2, s[0] == "sessions", method == "PUT" || method == "DELETE" {
            try patchSession(id: s[1], item, db)
        } else if method == "POST", s.count == 3, s[0] == "sessions", s[2] == "laps" {
            try appendLaps(sessionId: s[1], item, db)
        } else if method == "DELETE", s.count == 2, s[0] == "laps" {
            try deleteLap(id: s[1], db)
        } else if method == "PUT", s.count == 2, s[0] == "tracks" {
            try patchTrack(id: s[1], item, db)
        }
    }

    // MARK: - Events

    private static func createEvent(_ item: QueuedWrite, _ db: Database) throws {
        guard let tempId = item.tempId, let bodyData = item.body,
              let body = try? JSONDecoder().decode(EventBody.self, from: bodyData),
              let startDate = body.start_date
        else { return }

        // The track name is unknown offline unless the tracks list is cached; the
        // server resolves the real one (and the vehicle link) on flush.
        var trackName = body.track_name ?? "(new track)"
        if let trackId = body.track_id,
           let cached = try cached("/tracks", db),
           let tracks = try? JSONDecoder().decode([Track].self, from: cached),
           let track = tracks.first(where: { $0.id == trackId }) {
            trackName = track.name
        }

        var detail = EventDetail(
            event: Event(
                id: tempId,
                trackId: body.track_id ?? 0,
                trackName: trackName,
                startDate: startDate,
                days: body.days ?? 1,
                club: body.club,
                runGroup: body.run_group,
                car: body.car,
                vehicleId: nil,
                notes: body.notes,
                conditions: body.conditions,
                tempF: body.temp_f,
                checklist: body.checklist,
                bestTimeMs: body.best_time_ms,
                trackHours: body.track_hours,
                updatedAt: 0,
                lapBestMs: nil,
                lapCount: 0,
                sessionCount: 0,
                bestMs: nil,
                consistency: nil,
                hours: 0
            ),
            sessions: [],
            setups: []
        )
        OfflineMirrors.recomputeDetail(&detail)
        try put("/events/\(tempId)", try JSONEncoder().encode(detail), db)

        try patchEventLists(db) { list, key in
            // Track-filtered lists are left alone: offline, we may not know which
            // track this event belongs to.
            guard key == "/events" else { return nil }
            var rest = list.filter { $0.id != tempId }
            let at = rest.firstIndex { $0.startDate <= detail.event.startDate } ?? rest.count
            rest.insert(detail.event, at: at)
            return rest
        }
    }

    private static func updateEvent(id: String, _ item: QueuedWrite, _ db: Database) throws {
        guard let bodyData = item.body,
              let body = try? JSONDecoder().decode(EventBody.self, from: bodyData),
              let keys = try? JSONSerialization.jsonObject(with: bodyData) as? [String: Any]
        else { return }

        let path = "/events/\(id)"
        if let cached = try cached(path, db), var detail = try? JSONDecoder().decode(EventDetail.self, from: cached) {
            apply(body, keys: keys, to: &detail.event)
            OfflineMirrors.recomputeDetail(&detail)
            try put(path, try JSONEncoder().encode(detail), db)
            try syncLists(from: detail, db)
        } else {
            // No detail cached: patch the list rows directly so the dashboard still
            // shows the edit.
            try patchEventLists(db) { list, _ in
                list.map { row in
                    guard String(row.id) == id else { return row }
                    var next = row
                    apply(body, keys: keys, to: &next)
                    return next
                }
            }
        }
    }

    /// Only the fields the request actually carried — the server updates just those,
    /// so a patch that filled in the rest would invent data.
    private static func apply(_ body: EventBody, keys: [String: Any], to event: inout Event) {
        if keys.keys.contains("track_name"), let value = body.track_name { event.trackName = value }
        if keys.keys.contains("start_date"), let value = body.start_date { event.startDate = value }
        if keys.keys.contains("days"), let value = body.days { event.days = value }
        if keys.keys.contains("club") { event.club = body.club }
        if keys.keys.contains("run_group") { event.runGroup = body.run_group }
        if keys.keys.contains("car") { event.car = body.car }
        if keys.keys.contains("notes") { event.notes = body.notes }
        if keys.keys.contains("conditions") { event.conditions = body.conditions }
        if keys.keys.contains("temp_f") { event.tempF = body.temp_f }
        if keys.keys.contains("checklist") { event.checklist = body.checklist }
        if keys.keys.contains("best_time_ms") { event.bestTimeMs = body.best_time_ms }
        if keys.keys.contains("track_hours") { event.trackHours = body.track_hours }
    }

    private static func deleteEvent(id: String, _ db: Database) throws {
        try db.execute(sql: "DELETE FROM cached_response WHERE path = ?", arguments: ["/events/\(id)"])
        try patchEventLists(db) { list, _ in list.filter { String($0.id) != id } }
    }

    // MARK: - Setup sheets (inside the event detail)

    private static func patchSetup(eventId: String, day: String, _ item: QueuedWrite, _ db: Database) throws {
        let path = "/events/\(eventId)"
        guard let day = Int(day), let cached = try cached(path, db),
              var detail = try? JSONDecoder().decode(EventDetail.self, from: cached)
        else { return }

        var setups = detail.setups.filter { $0.day != day }
        if item.method == "PUT", let bodyData = item.body,
           let sheet = try? JSONDecoder().decode(SetupSheet.self, from: bodyData) {
            setups.append(Setup(day: day, data: sheet))
        }
        detail.setups = setups.sorted { $0.day < $1.day }
        try put(path, try JSONEncoder().encode(detail), db)
    }

    // MARK: - Sessions and laps

    private static func addSession(eventId: String, _ item: QueuedWrite, _ db: Database) throws {
        let path = "/events/\(eventId)"
        guard let tempId = item.tempId, let cached = try cached(path, db),
              var detail = try? JSONDecoder().decode(EventDetail.self, from: cached)
        else { return }
        // Re-apply dedupe.
        guard !detail.sessions.contains(where: { $0.id == tempId }) else { return }

        let body = item.body.flatMap { try? JSONDecoder().decode(SessionBody.self, from: $0) }
        let laps = OfflineMirrors.cleanLaps(body?.laps ?? [])
        detail.sessions.append(
            Session(
                id: tempId,
                label: body?.label,
                notes: body?.notes,
                sort: (detail.sessions.map(\.sort).max() ?? 0) + 1,
                trace: body?.trace,
                channels: body?.channels,
                laps: laps.enumerated().map { index, ms in
                    Lap(
                        id: tempLapId(seed: tempId, lapNum: index + 1),
                        sessionId: tempId,
                        lapNum: index + 1,
                        timeMs: ms
                    )
                }
            )
        )
        OfflineMirrors.recomputeDetail(&detail)
        try put(path, try JSONEncoder().encode(detail), db)
        try syncLists(from: detail, db)
    }

    private static func patchSession(id: String, _ item: QueuedWrite, _ db: Database) throws {
        try eachEventDetail(db) { path, detail in
            guard let index = detail.sessions.firstIndex(where: { String($0.id) == id }) else { return false }
            if item.method == "PUT" {
                // The route writes both columns unconditionally (missing → null),
                // so mirror that rather than merging.
                let body = item.body.flatMap { try? JSONDecoder().decode(SessionBody.self, from: $0) }
                detail.sessions[index].label = body?.label
                detail.sessions[index].notes = body?.notes
            } else {
                detail.sessions.remove(at: index)
            }
            return true
        }
    }

    private static func appendLaps(sessionId: String, _ item: QueuedWrite, _ db: Database) throws {
        guard let qid = item.qid.map(Int.init) else { return }
        let seed = qid + 100_000
        try eachEventDetail(db) { _, detail in
            guard let index = detail.sessions.firstIndex(where: { String($0.id) == sessionId }) else { return false }
            // Re-apply dedupe. Lap numbers continue from whatever the session
            // already had, so the check is "does any lap belong to this queue
            // entry?" — every id this entry mints falls in its own 1000-wide band,
            // which is the typed equivalent of the web mirror's `q{qid}-l` marker.
            let band = abs(seed) * 1000
            let alreadyApplied = detail.sessions[index].laps.contains { lap in
                lap.id < 0 && abs(lap.id) > band && abs(lap.id) < band + 1000
            }
            guard !alreadyApplied else { return true }

            var lapNum = detail.sessions[index].laps.map(\.lapNum).max() ?? 0
            let body = item.body.flatMap { try? JSONDecoder().decode(LapsBody.self, from: $0) }
            for ms in OfflineMirrors.cleanLaps(body?.laps ?? []) {
                lapNum += 1
                detail.sessions[index].laps.append(
                    Lap(
                        id: tempLapId(seed: seed, lapNum: lapNum),
                        sessionId: detail.sessions[index].id,
                        lapNum: lapNum,
                        timeMs: ms
                    )
                )
            }
            return true
        }
    }

    private static func deleteLap(id: String, _ db: Database) throws {
        try eachEventDetail(db) { _, detail in
            guard let index = detail.sessions.firstIndex(where: { session in
                session.laps.contains { String($0.id) == id }
            }) else { return false }
            detail.sessions[index].laps.removeAll { String($0.id) == id }
            return true
        }
    }

    // MARK: - Tracks

    private static func patchTrack(id: String, _ item: QueuedWrite, _ db: Database) throws {
        guard let bodyData = item.body,
              let body = try? JSONDecoder().decode(TrackBody.self, from: bodyData),
              let keys = try? JSONSerialization.jsonObject(with: bodyData) as? [String: Any],
              let cached = try cached("/tracks", db),
              let tracks = try? JSONDecoder().decode([Track].self, from: cached)
        else { return }

        let patched = tracks.map { track -> Track in
            guard String(track.id) == id else { return track }
            var next = track
            if keys.keys.contains("name"), let value = body.name { next.name = value }
            if keys.keys.contains("notes") { next.notes = body.notes }
            if keys.keys.contains("goal_ms") { next.goalMs = body.goal_ms }
            return next
        }
        try put("/tracks", try JSONEncoder().encode(patched), db)
    }

    // MARK: - Cache plumbing

    private static func cached(_ path: String, _ db: Database) throws -> Data? {
        try Data.fetchOne(db, sql: "SELECT body FROM cached_response WHERE path = ?", arguments: [path])
    }

    private static func put(_ path: String, _ body: Data, _ db: Database) throws {
        try db.execute(
            sql: "INSERT OR REPLACE INTO cached_response (path, body) VALUES (?, ?)",
            arguments: [path, body]
        )
    }

    /// Every cached events list — the unfiltered one and any track-filtered ones.
    private static func patchEventLists(_ db: Database, _ transform: ([Event], String) -> [Event]?) throws {
        let keys = try String.fetchAll(db, sql: "SELECT path FROM cached_response")
        for key in keys where key == "/events" || key.hasPrefix("/events?") {
            guard let data = try cached(key, db),
                  let list = try? JSONDecoder().decode([Event].self, from: data),
                  let next = transform(list, key)
            else { continue }
            try put(key, try JSONEncoder().encode(next), db)
        }
    }

    /// Keep the list rows in step with a patched detail: the same event appears in
    /// the dashboard list and on the track page, and they must agree.
    private static func syncLists(from detail: EventDetail, _ db: Database) throws {
        let row = OfflineMirrors.listRow(from: detail)
        try patchEventLists(db) { list, _ in
            list.map { $0.id == row.id ? row : $0 }
        }
    }

    /// Walk cached event details until `transform` claims one, then persist it and
    /// re-sync the lists. Sessions and laps are addressed by their own ids, so which
    /// event holds them isn't known from the path.
    private static func eachEventDetail(_ db: Database, _ transform: (String, inout EventDetail) -> Bool) throws {
        let keys = try String.fetchAll(db, sql: "SELECT path FROM cached_response")
        for key in keys {
            guard key.hasPrefix("/events/"), !key.contains("?"),
                  key.split(separator: "/").count == 2,
                  let data = try cached(key, db),
                  var detail = try? JSONDecoder().decode(EventDetail.self, from: data)
            else { continue }
            guard transform(key, &detail) else { continue }
            OfflineMirrors.recomputeDetail(&detail)
            try put(key, try JSONEncoder().encode(detail), db)
            try syncLists(from: detail, db)
            return
        }
    }
}

File: apps/android/core/src/main/kotlin/app/trackevolution/core/offline/OfflinePatch.kt
package app.trackevolution.core.offline

import app.trackevolution.core.model.ChecklistItem
import app.trackevolution.core.model.Conditions
import app.trackevolution.core.model.Event
import app.trackevolution.core.model.EventDetail
import app.trackevolution.core.model.Lap
import app.trackevolution.core.model.Session
import app.trackevolution.core.model.SessionChannels
import app.trackevolution.core.model.Setup
import app.trackevolution.core.model.SetupSheet
import app.trackevolution.core.model.Track
import app.trackevolution.core.model.TracePoint
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject

// Optimistic patching: a queued mutation edits the cached responses it affects, so
// the UI shows the change immediately and every view agrees — the dashboard, the
// event page and the track page are three different cache entries describing the
// same rows.
//
// Patches are best-effort *display* state. A missing cache entry is skipped, never
// created: the queue is the source of truth until it flushes. Every patch is
// idempotent, because `reapplyQueue` runs them again over a fresh server response.
//
// Port of `applyLocal` in `public/js/offline.js`, function for function. Unlike
// that mirror, which surgically edits plain JSON, this decodes into the typed
// models and re-encodes — see `OfflineJson` for why that is safe.

/**
 * Apply one queued mutation to the cache, inside the caller's transaction.
 *
 * A path this doesn't recognise is simply not mirrored; the queueable whitelist
 * in [OfflineStore] is what decides whether it should have been queued at all.
 */
internal suspend fun applyLocal(item: QueuedWrite, txn: OfflineTransaction) {
    val s = item.path.split("/").filter { it.isNotEmpty() }
    val method = item.method
    when {
        method == "POST" && s == listOf("events") ->
            createEvent(item, txn)

        method == "PUT" && s.size == 2 && s[0] == "events" ->
            updateEvent(s[1], item, txn)

        method == "DELETE" && s.size == 2 && s[0] == "events" ->
            deleteEvent(s[1], txn)

        s.size == 4 && s[0] == "events" && s[2] == "setups" && (method == "PUT" || method == "DELETE") ->
            patchSetup(s[1], s[3], item, txn)

        method == "POST" && s.size == 3 && s[0] == "events" && s[2] == "sessions" ->
            addSession(s[1], item, txn)

        s.size == 2 && s[0] == "sessions" && (method == "PUT" || method == "DELETE") ->
            patchSession(s[1], item, txn)

        method == "POST" && s.size == 3 && s[0] == "sessions" && s[2] == "laps" ->
            appendLaps(s[1], item, txn)

        method == "DELETE" && s.size == 2 && s[0] == "laps" ->
            deleteLap(s[1], txn)

        method == "PUT" && s.size == 2 && s[0] == "tracks" ->
            patchTrack(s[1], item, txn)
    }
}

// ---- Events ---------------------------------------------------------------

private suspend fun createEvent(item: QueuedWrite, txn: OfflineTransaction) {
    val tempId = item.tempId ?: return
    val body = item.bodyObject() ?: return
    val startDate = body.string("start_date") ?: return
    val trackId = body.int("track_id")

    // The track name is unknown offline unless the tracks list is cached; the
    // server resolves the real one (and the vehicle link) on flush.
    val trackName = trackId?.let { id ->
        txn.tracks()?.firstOrNull { it.id == id }?.name
    } ?: body.string("track_name") ?: "(new track)"

    val detail = OfflineMirrors.recomputeDetail(
        EventDetail(
            event = Event(
                id = tempId,
                trackId = trackId ?: 0,
                trackName = trackName,
                startDate = startDate,
                days = body.double("days") ?: 1.0,
                club = body.string("club"),
                runGroup = body.string("run_group"),
                car = body.string("car"),
                // Matched server-side from the car name on flush.
                vehicleId = null,
                notes = body.string("notes"),
                conditions = body.string("conditions")?.let(::Conditions),
                tempF = body.int("temp_f"),
                checklist = body.decode("checklist", checklistSerializer),
                bestTimeMs = body.int("best_time_ms"),
                trackHours = body.double("track_hours"),
                updatedAt = 0,
                lapBestMs = null,
                lapCount = 0,
                sessionCount = 0,
                bestMs = null,
                consistency = null,
                hours = 0.0,
            ),
            sessions = emptyList(),
            setups = emptyList(),
        ),
    )

    txn.putDetail("/events/$tempId", detail)
    txn.patchEventLists { list, key ->
        // Track-filtered lists are left alone: offline, we may not know which
        // track this event belongs to.
        if (key != "/events") return@patchEventLists null
        val rest = list.filter { it.id != tempId }.toMutableList()
        val at = rest.indexOfFirst { it.startDate <= detail.event.startDate }
        rest.add(if (at == -1) rest.size else at, detail.event)
        rest
    }
}

private suspend fun updateEvent(id: String, item: QueuedWrite, txn: OfflineTransaction) {
    val body = item.bodyObject() ?: return
    val path = "/events/$id"
    val cached = txn.detail(path)
    if (cached != null) {
        val patched = OfflineMirrors.recomputeDetail(cached.copy(event = cached.event.patched(body)))
        txn.putDetail(path, patched)
        txn.syncLists(patched)
    } else {
        // No detail cached: patch the list rows directly so the dashboard still
        // shows the edit.
        txn.patchEventLists { list, _ ->
            list.map { row -> if (row.id.toString() == id) row.patched(body) else row }
        }
    }
}

/**
 * Only the fields the request actually carried — the server updates just those
 * (`if ("notes" in body)` in `src/routes/events.ts`), so a patch that filled in
 * the rest would invent data.
 */
private fun Event.patched(body: JsonObject): Event {
    var next = this
    if (body.has("track_name")) body.string("track_name")?.let { next = next.copy(trackName = it) }
    if (body.has("start_date")) body.string("start_date")?.let { next = next.copy(startDate = it) }
    if (body.has("days")) body.double("days")?.let { next = next.copy(days = it) }
    if (body.has("club")) next = next.copy(club = body.string("club"))
    if (body.has("run_group")) next = next.copy(runGroup = body.string("run_group"))
    if (body.has("car")) next = next.copy(car = body.string("car"))
    if (body.has("notes")) next = next.copy(notes = body.string("notes"))
    if (body.has("conditions")) {
        next = next.copy(conditions = body.string("conditions")?.let(::Conditions))
    }
    if (body.has("temp_f")) next = next.copy(tempF = body.int("temp_f"))
    if (body.has("checklist")) {
        next = next.copy(checklist = body.decode("checklist", checklistSerializer))
    }
    if (body.has("best_time_ms")) next = next.copy(bestTimeMs = body.int("best_time_ms"))
    if (body.has("track_hours")) next = next.copy(trackHours = body.double("track_hours"))
    return next
}

private suspend fun deleteEvent(id: String, txn: OfflineTransaction) {
    txn.responseDelete("/events/$id")
    txn.patchEventLists { list, _ -> list.filter { it.id.toString() != id } }
}

// ---- Setup sheets (inside the event detail) -------------------------------

private suspend fun patchSetup(
    eventId: String,
    day: String,
    item: QueuedWrite,
    txn: OfflineTransaction,
) {
    val dayNumber = day.toIntOrNull() ?: return
    val path = "/events/$eventId"
    val detail = txn.detail(path) ?: return

    val setups = detail.setups.filterNot { it.day == dayNumber }.toMutableList()
    if (item.method == "PUT") {
        val sheet = item.body?.let {
            runCatching { OfflineJson.decode.decodeFromString(SetupSheet.serializer(), it) }.getOrNull()
        }
        if (sheet != null) setups.add(Setup(day = dayNumber, data = sheet))
    }
    txn.putDetail(path, detail.copy(setups = setups.sortedBy { it.day }))
}

// ---- Sessions and laps ----------------------------------------------------

private suspend fun addSession(eventId: String, item: QueuedWrite, txn: OfflineTransaction) {
    val tempId = item.tempId ?: return
    val path = "/events/$eventId"
    val detail = txn.detail(path) ?: return
    // Re-apply dedupe.
    if (detail.sessions.any { it.id == tempId }) return

    val body = item.bodyObject()
    val laps = OfflineMirrors.cleanLaps(body.doubleList("laps"))
    val session = Session(
        id = tempId,
        label = body?.string("label"),
        notes = body?.string("notes"),
        sort = (detail.sessions.maxOfOrNull { it.sort } ?: 0) + 1,
        trace = body?.decode("trace", traceSerializer),
        channels = body?.decode("channels", SessionChannels.serializer()),
        laps = laps.mapIndexed { index, ms ->
            Lap(
                id = tempLapId(seed = tempId, lapNum = index + 1),
                sessionId = tempId,
                lapNum = index + 1,
                timeMs = ms,
            )
        },
    )

    val patched = OfflineMirrors.recomputeDetail(detail.copy(sessions = detail.sessions + session))
    txn.putDetail(path, patched)
    txn.syncLists(patched)
}

private suspend fun patchSession(id: String, item: QueuedWrite, txn: OfflineTransaction) {
    txn.eachEventDetail { detail ->
        val index = detail.sessions.indexOfFirst { it.id.toString() == id }
        if (index == -1) return@eachEventDetail null
        val sessions = detail.sessions.toMutableList()
        if (item.method == "PUT") {
            // The route writes both columns unconditionally (`body.label ?? null`),
            // so mirror that rather than merging.
            val body = item.bodyObject()
            sessions[index] = sessions[index].copy(
                label = body?.string("label"),
                notes = body?.string("notes"),
            )
        } else {
            sessions.removeAt(index)
        }
        detail.copy(sessions = sessions)
    }
}

private suspend fun appendLaps(sessionId: String, item: QueuedWrite, txn: OfflineTransaction) {
    // Seeded off the queue id, and offset clear of the temp-id range, so a
    // re-apply mints the same lap ids and dedupes instead of duplicating.
    val seed = item.qid.toInt() + LAP_SEED_OFFSET
    txn.eachEventDetail { detail ->
        val index = detail.sessions.indexOfFirst { it.id.toString() == sessionId }
        if (index == -1) return@eachEventDetail null

        // Re-apply dedupe. Lap numbers continue from whatever the session already
        // had, so the question is "does any lap belong to this queue entry?" —
        // every id this entry mints falls in its own 1000-wide band, which is the
        // typed equivalent of the web mirror's `q{qid}-l` marker.
        val band = seed * LAP_BAND
        val session = detail.sessions[index]
        val alreadyApplied = session.laps.any { it.id < 0 && -it.id > band && -it.id < band + LAP_BAND }
        if (alreadyApplied) return@eachEventDetail detail

        var lapNum = session.laps.maxOfOrNull { it.lapNum } ?: 0
        val added = OfflineMirrors.cleanLaps(item.bodyObject().doubleList("laps")).map { ms ->
            lapNum += 1
            Lap(
                id = tempLapId(seed = seed, lapNum = lapNum),
                sessionId = session.id,
                lapNum = lapNum,
                timeMs = ms,
            )
        }
        val sessions = detail.sessions.toMutableList()
        sessions[index] = session.copy(laps = session.laps + added)
        detail.copy(sessions = sessions)
    }
}

private suspend fun deleteLap(id: String, txn: OfflineTransaction) {
    txn.eachEventDetail { detail ->
        val index = detail.sessions.indexOfFirst { session ->
            session.laps.any { it.id.toString() == id }
        }
        if (index == -1) return@eachEventDetail null
        val sessions = detail.sessions.toMutableList()
        sessions[index] = sessions[index].copy(
            laps = sessions[index].laps.filterNot { it.id.toString() == id },
        )
        detail.copy(sessions = sessions)
    }
}

/** Local ids for rows the server hasn't seen; negative, like every temp id. */
private fun tempLapId(seed: Int, lapNum: Int): Int = -(kotlin.math.abs(seed) * LAP_BAND + lapNum)

private const val LAP_BAND = 1000

/** Keeps appended-lap seeds clear of the session temp ids, which start at 1. */
private const val LAP_SEED_OFFSET = 100_000

// ---- Tracks ---------------------------------------------------------------

private suspend fun patchTrack(id: String, item: QueuedWrite, txn: OfflineTransaction) {
    val body = item.bodyObject() ?: return
    val tracks = txn.tracks() ?: return
    val patched = tracks.map { track ->
        if (track.id.toString() != id) {
            track
        } else {
            var next = track
            if (body.has("name")) body.string("name")?.let { next = next.copy(name = it) }
            if (body.has("notes")) next = next.copy(notes = body.string("notes"))
            if (body.has("goal_ms")) next = next.copy(goalMs = body.int("goal_ms"))
            next
        }
    }
    txn.responsePut("/tracks", OfflineJson.encode.encodeToString(trackListSerializer, patched))
}

// ---- Cache plumbing -------------------------------------------------------

private suspend fun OfflineTransaction.detail(path: String): EventDetail? {
    val raw = responseGet(path) ?: return null
    return runCatching { OfflineJson.decode.decodeFromString(EventDetail.serializer(), raw) }.getOrNull()
}

private suspend fun OfflineTransaction.putDetail(path: String, detail: EventDetail) {
    responsePut(path, OfflineJson.encode.encodeToString(EventDetail.serializer(), detail))
}

private suspend fun OfflineTransaction.tracks(): List<Track>? {
    val raw = responseGet("/tracks") ?: return null
    return runCatching { OfflineJson.decode.decodeFromString(trackListSerializer, raw) }.getOrNull()
}

/** Every cached events list — the unfiltered one and any track-filtered ones. */
private suspend fun OfflineTransaction.patchEventLists(
    transform: (List<Event>, String) -> List<Event>?,
) {
    for (key in responsePaths()) {
        if (key != "/events" && !key.startsWith("/events?")) continue
        val raw = responseGet(key) ?: continue
        val list = runCatching { OfflineJson.decode.decodeFromString(eventListSerializer, raw) }
            .getOrNull() ?: continue
        val next = transform(list, key) ?: continue
        responsePut(key, OfflineJson.encode.encodeToString(eventListSerializer, next))
    }
}

/**
 * Keep the list rows in step with a patched detail: the same event appears in the
 * dashboard list and on the track page, and they must agree.
 */
private suspend fun OfflineTransaction.syncLists(detail: EventDetail) {
    val row = OfflineMirrors.listRow(detail)
    patchEventLists { list, _ -> list.map { if (it.id == row.id) row else it } }
}

/**
 * Walk cached event details until [transform] claims one, then recompute it,
 * persist it, and re-sync the lists. Sessions and laps are addressed by their own
 * ids, so which event holds them isn't knowable from the path.
 */
private suspend fun OfflineTransaction.eachEventDetail(
    transform: (EventDetail) -> EventDetail?,
) {
    for (key in responsePaths()) {
        if (!key.startsWith("/events/") || key.contains("?")) continue
        if (key.split("/").filter { it.isNotEmpty() }.size != 2) continue
        val detail = detail(key) ?: continue
        val next = transform(detail) ?: continue
        val recomputed = OfflineMirrors.recomputeDetail(next)
        putDetail(key, recomputed)
        syncLists(recomputed)
        return
    }
}

// ---- Request-body reading -------------------------------------------------
//
// A queued body is the raw request JSON, so "was this key present?" — the
// distinction the server's update routes turn on — is answerable directly, which
// is a good deal simpler than the iOS port, where it takes a second parse.

private fun QueuedWrite.bodyObject(): JsonObject? = body?.let {
    runCatching { OfflineJson.decode.parseToJsonElement(it).jsonObject }.getOrNull()
}

private fun JsonObject.has(key: String): Boolean = containsKey(key)

private fun JsonObject.value(key: String): JsonElement? = this[key]?.takeIf { it !is JsonNull }

private fun JsonObject.string(key: String): String? =
    (value(key) as? JsonPrimitive)?.takeIf { it.isString }?.content

private fun JsonObject.int(key: String): Int? = (value(key) as? JsonPrimitive)?.intOrNull

private fun JsonObject.double(key: String): Double? = (value(key) as? JsonPrimitive)?.doubleOrNull

private fun <T> JsonObject.decode(
    key: String,
    serializer: kotlinx.serialization.KSerializer<T>,
): T? = value(key)?.let {
    runCatching { OfflineJson.decode.decodeFromJsonElement(serializer, it) }.getOrNull()
}

/**
 * Lap times as sent, before [OfflineMirrors.cleanLaps] gets them: anything that
 * isn't a number becomes NaN and is dropped there, exactly as `Number(v)` does in
 * the web mirror.
 */
private fun JsonObject?.doubleList(key: String): List<Double> {
    val array = this?.value(key) as? JsonArray ?: return emptyList()
    return array.map { element -> (element as? JsonPrimitive)?.doubleOrNull ?: Double.NaN }
}

private val checklistSerializer = ListSerializer(ChecklistItem.serializer())
private val traceSerializer = ListSerializer(TracePoint.serializer())
private val eventListSerializer = ListSerializer(Event.serializer())
private val trackListSerializer = ListSerializer(Track.serializer())
